### PR TITLE
Add flatpak json schema

### PR DESCRIPTION
**Pull request by Hofer-Julian**
Tasks to be finished:

- [x] Finish schema

- [x] [Map schema in the extension](https://code.visualstudio.com/api/references/contribution-points#_contributes.jsonValidation)

- [x] Add [extension dependency](https://code.visualstudio.com/api/references/extension-manifest#fields) on [YAML](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml)

- [x] Check if flatpak manifests work out of the box (yaml, yml and json)

### CODE REVIEW

**Comment by Hofer-Julian** [comment]
@bilelmoussaoui the schema is finished so far, if you want to you can play a little bit with manifests. At least with the approx. 20 I've tested everything seems to work fine. 

I've noticed one bug with the yaml extension though, which would be interesting to know if you can reproduce it:
- open a yaml file
- hover of the content of one source
- "type" will give output, but the others will not
- if you make a new line, intellisense(ctrl+space) suggest the correct options and for these hovering then works as well

Additionally, I wonder if the schema could be used to replace the current way of manifest detection.

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui the schema is finished so far, if you want to you can play a little bit with manifests. At least with the approx. 20 I've tested everything seems to work fine.
> 
> I've noticed one bug with the yaml extension though, which would be interesting to know if you can reproduce it:
> 
>     * open a yaml file
> 
>     * hover of the content of one source
> 
>     * "type" will give output, but the others will not
> 
>     * if you make a new line, intellisense(ctrl+space) suggest the correct options and for these hovering then works as well
> 
> 
> Additionally, I wonder if the schema could be used to replace the current way of manifest detection.

Thanks! I will give it a try. Regarding the detection of the schema, I don't think there's any api to do so... there's probably some nodejs library for that though. I wonder as well in case it's not possible to do that the other way around, to be able to set the schema only for files that were detected as a flatpak manifest?

File: .vscode/settings.json [hunk 1]
```diff
@@ -15,5 +15,25 @@
     "devel",
     "freedesktop",
     "nofilesystem"
-  ]
-}
+  ],
+  "json.schemas": [
```
Comment (bilelmoussaoui) [inline .vscode/settings.json:19]: This is not correct, it should be defined in package.json instead otherwise it won't work

Comment (bilelmoussaoui) [reply .vscode/settings.json:19]:
```json

    "jsonValidation": [
      {
        "fileMatch": [
          "/*.*.*.json",
          "/*.*.*.*.json",
          "/*.*.*.*.*.json"
        ],
        "url": "./myschema.json"
      }
    ]
```
something like this works for the json, no idea how to setup the yaml one though

Comment (Hofer-Julian) [reply .vscode/settings.json:19]: That‘s why it‘s a draft.
I commited it so that it is easier for you to test 😉

**Comment by Hofer-Julian** [comment]
> > @bilelmoussaoui the schema is finished so far, if you want to you can play a little bit with manifests. At least with the approx. 20 I've tested everything seems to work fine.
> > I've noticed one bug with the yaml extension though, which would be interesting to know if you can reproduce it:
> > ```
> > * open a yaml file
> > 
> > * hover of the content of one source
> > 
> > * "type" will give output, but the others will not
> > 
> > * if you make a new line, intellisense(ctrl+space) suggest the correct options and for these hovering then works as well
> > ```
> > 
> > 
> > Additionally, I wonder if the schema could be used to replace the current way of manifest detection.
> 
> Thanks! I will give it a try. Regarding the detection of the schema, I don't think there's any api to do so... there's probably some nodejs library for that though. I wonder as well in case it's not possible to do that the other way around, to be able to set the schema only for files that were detected as a flatpak manifest?

Yeah, I assume some dependency would be necessary.
For the reverse case, I couldn‘t find a possibility, but maybe you are more successful

File: .vscode/settings.json (hunk 1 above)
Comment (bilelmoussaoui) [reply .vscode/settings.json:19]: Sure, I just mentioned it because i didn't notice it was using the settings.json file. It works fine btw ^^

**Comment by bilelmoussaoui** [comment]
> > > @bilelmoussaoui the schema is finished so far, if you want to you can play a little bit with manifests. At least with the approx. 20 I've tested everything seems to work fine.
> > > I've noticed one bug with the yaml extension though, which would be interesting to know if you can reproduce it:
> > > ```
> > > * open a yaml file
> > > 
> > > * hover of the content of one source
> > > 
> > > * "type" will give output, but the others will not
> > > 
> > > * if you make a new line, intellisense(ctrl+space) suggest the correct options and for these hovering then works as well
> > > ```
> > > 
> > > 
> > > Additionally, I wonder if the schema could be used to replace the current way of manifest detection.
> > 
> > 
> > Thanks! I will give it a try. Regarding the detection of the schema, I don't think there's any api to do so... there's probably some nodejs library for that though. I wonder as well in case it's not possible to do that the other way around, to be able to set the schema only for files that were detected as a flatpak manifest?
> 
> Yeah, I assume some dependency would be necessary.
> For the reverse case, I couldn‘t find a possibility, but maybe you are more successful

I will have a look, that would be a better & more precise way to detect the manifest

File: .vscode/settings.json (hunk 1 above)
Comment (Hofer-Julian) [reply .vscode/settings.json:19]: Glad to hear that

File: package.json [hunk 2]
```diff
@@ -117,6 +117,42 @@
...
+        "url": "./manifest-schema.json"
+      },
+      {
+        "fileMatch": "/*.*.*.*.yml",
+        "url": "./manifest-schema.json"
+      },
+      {
+        "fileMatch": "/*.*.*.*.*.yml",
+        "url": "./manifest-schema.json"
+      }
```
Comment (bilelmoussaoui) [inline package.json:155]: Can't this be made an array like the json one?

File: package.json [hunk 3]
```diff
@@ -144,5 +180,8 @@
     "@types/js-yaml": "^3.12.5",
     "effector": "^21.6.0",
     "js-yaml": "^3.14.0"
-  }
-}
+  },
+  "extensionDependencies": [
+    "redhat.vscode-yaml"
+  ]
+}
```
Comment (bilelmoussaoui) [inline package.json:187]: missing new line at the end of the file

File: .vscode/settings.json [hunk 4]
```diff
@@ -16,4 +16,4 @@
     "freedesktop",
     "nofilesystem"
   ]
-}
+}
```
Comment (bilelmoussaoui) [inline .vscode/settings.json:19]: missing new line at the end of the file

File: package.json (hunk 2 above)
Comment (Hofer-Julian) [reply package.json:155]: Unfortunately not (at least it didn't work on my machine)
Probably also something for a feature request.

File: package-lock.json [hunk 5]
```diff
@@ -0,0 +1,1814 @@
+{
```
Comment (bilelmoussaoui) [inline package-lock.json:1]: Don't commit the package-lock.json, preferably would add it to .gitignore

File: manifest-schema.json [hunk 6]
```diff
@@ -0,0 +1,1022 @@
...
+            "type": "string"
+        },
+        "desktop-file-name-suffix": {
+            "description": "This string will be suffixed to the Name key in the main application desktop file.",
+            "type": "string"
+        },
+        "additionalProperties": false
+    },
+    "additionalProperties": false
+}
```
Comment (bilelmoussaoui) [inline manifest-schema.json:1022]: missing new line at the end of the file

File: manifest-schema.json [hunk 7]
```diff
@@ -0,0 +1,1022 @@
...
+        "app-id": {
+            "description": "A string defining the application id.",
+            "type": "string"
+        },
+        "id": {
+            "description": "A string defining the application id.",
+            "type": "string"
+        },
+        "branch": {
+            "description": "The default branch to use when exporting the application. Defaults to master. This key can be overriden by the --default-branch commandline option.",
```
Comment (bilelmoussaoui) [inline manifest-schema.json:810]:
```suggestion
            "description": "The default branch to use when exporting the application. Defaults to master. This key can be overriden by the --default-branch commandline option.",
```
```suggestion
            "description": "The default branch to use when exporting the application. Defaults to master. This key can be overridden by the --default-branch commandline option.",
```

File: package.json (hunk 2 above)
Comment (bilelmoussaoui) [reply package.json:155]: Ah, no worries then

File: package-lock.json (hunk 5 above)
Comment (Hofer-Julian) [reply package-lock.json]: Fixed

File: package.json (hunk 3 above)
Comment (Hofer-Julian) [reply package.json]: Done

File: manifest-schema.json (hunk 6 above)
Comment (Hofer-Julian) [reply manifest-schema.json]: Done

File: .vscode/settings.json (hunk 4 above)
Comment (Hofer-Julian) [reply .vscode/settings.json]: Done

**Comment by Hofer-Julian** [comment]
The PR can be merged now from my side.
You might want to squash it though

**Comment by bilelmoussaoui** [comment]
Seems to be working fine, awesome contribution! thanks a lot :D

**Pull request merged**